### ISSUE TITLE
anti-meridian crossing issue when reading raster with rasterio

### ISSUE THREAD

**Issue by Evelyn-M** [CLIMADA-project/climada_python#172]
Several of our functions which involve reading rasters into memory offer a bounding box or a geometry option for reading in only a part of the file. 
Currently, this breaks down when the anti-meridian is crossed (i.e. crossing from -179.9 to 179.9) in all functions that somehow use this. 
Examples:
```
bbox = (-179.5,-0.5,179.5,0.5)
geometry = [shapely.geometry.box(*bbox, ccw=True)]
# a raster file with resolution 0.0083333, i.e. one degree should be split into 121 units
path_sourcefile = '/Users/evelynm/climada_python/data/system/ls_pr/ls_pr.tif'


# example which uses geometry object (via rasterio.mask)
meta, inten = u_coord.read_raster(path_sourcefile, geometry=geometry)
# shape of resulting file: width: 43081 instead of 121

# example which uses tuple object (via simple substraction)
inten, meta = u_coord.read_raster_bounds(path_sourcefile, bounds=bbox)
# shape of resulting file: width: 43081 instead of 121


res=0.00833333333333333
self.centroids.set_raster_from_pnt_bounds(bbox,res,crs='EPSG:4326')
# shape of centroids: 43081x121 instead of 121x1221
```

Reasons are two-fold: some functions are manually coded (x_max - x_min), others use the rasterio.mask.mask(). Neither works in that case.

**Comment by tovogt** [comment]
The area from -179.9 to 179.9 is NOT crossing the antimeridian. Maybe you rather want to use `bbox = (179.5, -0.5, 180.5, 0.5)` or `bbox = (-180.5, -0.5, -179.5, 0.5)`. This is supported by `read_raster_bounds` at least.

**Comment by Evelyn-M** [comment]
ok maybe I'm using the word anti-meridian wrongly then. But what is this here:
<img width="628" alt="Screenshot 2021-03-18 at 16 53 28" src="https://user-images.githubusercontent.com/42063186/111656535-e94b2a80-880a-11eb-9e8a-cc9fe8c76478.png">
<img width="518" alt="Screenshot 2021-03-18 at 16 54 26" src="https://user-images.githubusercontent.com/42063186/111656546-eb14ee00-880a-11eb-8f81-66622eb34c9e.png">

**Comment by Evelyn-M** [comment]
So for that case, if i wanted a 1° bounding box that starts a bit west of fiji and goes a bit to the east of fiji, I'd have to put -179.5 to 179.5, which evaluates to 359 degrees in the implementation.

**Comment by tovogt** [comment]
Yes, the area from -179.5 to 179.5 has a width of 359 degrees. You want the area between 179.5 and 180.5. That's what I already wrote above:
> Maybe you rather want to use bbox = (179.5, -0.5, 180.5, 0.5) or bbox = (-180.5, -0.5, -179.5, 0.5)

**Comment by Evelyn-M** [comment]
I understood it's rather a projection thing than a fact. I see, the workaround definitely works, but it's then not necessarily consistent with the EPSG 4326 definition, which goes from -180 to +180. Maybe worth putting a not somewhere at least, then?

**Comment by chahank** [comment]
Another way would be to invert the values `bbox = (179.5, -0.5, -179.5, 0.5)`. There needs to an indicated direction (basically, we go from left to right).

**Comment by tovogt** [comment]
No, the EPSG:4326 definition doesn't exclude any longitudinal values, neither does WGS 84 (https://apps.dtic.mil/dtic/tr/fulltext/u2/a280358.pdf). Since it's a geodetic coordinate system, it just defines the ellipsoid and the 0 degree lines. Ellipsoidal coordinates are periodic in longitude. You can confirm this yourself by using the transform-tool on the EPSG website https://epsg.io/transform#s_srs=4326&t_srs=3857&x=180.5000000&y=0.0000000

More precisely, the coordinate system of EPSG:4326 is "Ellipsoidal 2D CS. Axes: latitude, longitude. Orientations: north, east. UoM: degree" (EPSG:6422) and this means: "Any degree representation (e.g. DMSH, decimal, etc.) may be used".

Note that the "WGS84 bounds:" specification on https://epsg.io/4326 just explains which area is covered by this coordinate system. It doesn't specifiy the admissible coordinate range.

**Comment by Evelyn-M** [comment]
> Another way would be to invert the values bbox = (179.5, -0.5, -179.5, 0.5)

Nope that does not work. That of course evaluates to 359 as well.
Plus a bbox is defined as (x_min, y_min, x_max, y_max), so that would be against convention

**Comment by tovogt** [comment]
Inverting the lower and upper bounds (`bbox = (179.5, -0.5, -179.5, 0.5)`) as suggested by @chahank is NOT a good solution, because the function `read_raster_bounds` is CRS-agnostic and needs to assume that the lower bound is really lower than the upper bound.

**Comment by tovogt** [comment]
For most use cases, you will want to normalize longitudinal coordinates using `u_coord.lon_normalize` after determining your central meridian using `u_coord.lon_bounds`.

**Comment by tovogt** [comment]
It's true, however, that `read_raster` (and, therefore, `Centroids.set_raster_file`) is not able to deal with geometries that cross the antimeridian. Maybe we can take this as the actual topic of this thread :)

**Comment by chahank** [comment]
@ValentinGebhart @DahyannAraya  maybe relevant for the bounding box plans?
